canonical/mysql-operator
# Refactor MySQL users creations + implement cluster creation and addition of instances to cluster

[Jira ticket 1](https://warthogs.atlassian.net/jira/software/c/projects/DPE/boards/390?modal=detail&selectedIssue=DPE-144) and [Jira ticket 2](https://warthogs.atlassian.net/jira/software/c/projects/DPE/boards/390?modal=detail&selectedIssue=DPE-145)

# Issue
1. We are not creating our users for mysql correctly. Additionally, we are only revoking a subset of the required priveleges for the `root` user
2. MySQL is binding to `127.0.0.1` upon installation of `mysql-server-8.0`. However, InnoDB requires cluster instances to be bound to `0.0.0.0`
3. We have not yet implemented the helper method to create a cluster
4. We have not yet implemented the helper method to add an instance to the cluster

# Solution
1. MySQL installation creates `root@localhost`. We additionally create `root@%` and `serverconfig@%` manually. We then create `clusteradmin@%` while configuring the instance for InnoDB (by calling `dba.configure_instance()` within a `mysqlsh` context). Additionally, we will revoke all system user related privileges from `root@localhost` and `root@%`.
See [this link](https://docs.oracle.com/cd/E17952_01/mysql-shell-8.0-en/innodb-cluster-user-accounts.html) for more information on InnoDB users. Also see [this link](https://dev.mysql.com/doc/refman/8.0/en/privileges-provided.html#priv_super) for more information on system user related privileges.

2. Add a mysql configuration template file (`templates/mysqld.cnf`) which defines values for `bind-address` and `mysqlx-bind-address` set to `0.0.0.0`. Copy over these files to the mysql configuration directory (with prefix `z-` so the file has the highest priority) and restart mysql via systemctl during the `on_install` hook.

3. Implement the helper method to create a cluster
4. Implement the helper method to add an instance to a cluster


# Testing
To test, I did the following:
1. Run `charmcraft pack` and deploy a unit using lxd. SSH into the lxd container and ensure that MySQL is running and confirm using the `mysql` client that the bind address values are set to `0.0.0.0`
```
shayan@wellington:~/code/mysql-operator (fix/refactor_user_creation) $ lxc exec juju-30620b-21 bash 
root@juju-30620b-21:~# systemctl status mysql
● mysql.service - MySQL Community Server
     Loaded: loaded (/lib/systemd/system/mysql.service; enabled; vendor preset: enabled)
     Active: active (running) since Thu 2022-04-07 20:11:51 UTC; 35s ago
    Process: 5816 ExecStartPre=/usr/share/mysql/mysql-systemd-start pre (code=exited, status=0/SUCCESS)
   Main PID: 5824 (mysqld)
     Status: "Server is operational"
      Tasks: 38 (limit: 18730)
     Memory: 357.9M
     CGroup: /system.slice/mysql.service
             └─5824 /usr/sbin/mysqld

Apr 07 20:11:50 juju-30620b-21 systemd[1]: Starting MySQL Community Server...
Apr 07 20:11:51 juju-30620b-21 systemd[1]: Started MySQL Community Server.
root@juju-30620b-21:~# mysql

mysql> show variables like '%addr%';
+---------------------+---------+
| Variable_name       | Value   |
+---------------------+---------+
| admin_address       |         |
| bind_address        | 0.0.0.0 |
| mysqlx_bind_address | 0.0.0.0 |
+---------------------+---------+
3 rows in set (0.00 sec)
```

2. I launched 2 units with juju using lxd. Then SSH into both units, navigate to the juju source code directory, instantiate an instance of `MySQL` helper class and run the `configure_mysql_users` and `configure_instance` on both lxd containers. Then, on one container run `create_cluster()` and `add_instance_to_cluster()` methods.

# Release Notes
- Implement the helper functions to configure mysql, create the appropriate mysql users, configure the instance, create a cluster and add instances to the cluster.
